File: libmuscle/cpp/src/libmuscle/tests/test_peer_info.cpp
#include "libmuscle/peer_info.hpp"

#include <libmuscle/namespace.hpp>
#include <ymmsl/ymmsl.hpp>

#include <memory>

#include <gtest/gtest.h>


using libmuscle::_MUSCLE_IMPL_NS::PeerDims;
using libmuscle::_MUSCLE_IMPL_NS::PeerLocations;
using libmuscle::_MUSCLE_IMPL_NS::PeerInfo;
using ymmsl::Conduit;
using ymmsl::Reference;


int main(int argc, char *argv[]) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}


PeerInfo peer_info() {
    Reference kernel("kernel");
    std::vector<int> index({13});

    std::vector<Conduit> conduits({
        Conduit("kernel.out", "other.in"),
        Conduit("other.out", "kernel.in")});

    PeerDims peer_dims({{Reference("other"), {}}});

    PeerLocations peer_locations({
            {Reference("other"), {"tcp:other"}}});

    return PeerInfo(kernel, index, conduits, peer_dims, peer_locations);
}

PeerInfo peer_info2() {
    Reference kernel("other");
    std::vector<int> index;

    std::vector<Conduit> conduits({
        Conduit("kernel.out", "other.in"),
        Conduit("other.out", "kernel.in")});

    PeerDims peer_dims({{Reference("kernel"), {20}}});

    PeerLocations peer_locations({
            {Reference("kernel"), {"tcp:kernel"}}});

    return PeerInfo(kernel, index, conduits, peer_dims, peer_locations);
}

PeerInfo peer_info3() {
    Reference kernel("kernel");
    std::vector<int> index;

    std::vector<Conduit> conduits({
        Conduit("kernel.out", "other.in"),
        Conduit("other.out", "kernel.in")});

    PeerDims peer_dims({{Reference("other"), {}}});

    PeerLocations peer_locations({
            {Reference("other"), {"tcp:other"}}});

    return PeerInfo(kernel, index, conduits, peer_dims, peer_locations);
}

TEST(libmuscle_peer_info, create_peer_info) {
    peer_info();
}

TEST(libmuscle_peer_info, is_connected) {
    auto pi = peer_info();

    ASSERT_TRUE(pi.is_connected("out"));
    ASSERT_TRUE(pi.is_connected("in"));
    ASSERT_FALSE(pi.is_connected("not_connected"));
}

TEST(libmuscle_peer_info, get_peer_port) {
    auto pi = peer_info();
    ASSERT_EQ(pi.get_peer_ports("out"), std::vector<Reference>({"other.in"}));
    ASSERT_EQ(pi.get_peer_ports("in"), std::vector<Reference>({"other.out"}));

    auto pi2 = peer_info2();
    ASSERT_EQ(pi2.get_peer_ports("out"), std::vector<Reference>({"kernel.in"}));
    ASSERT_EQ(pi2.get_peer_ports("in"), std::vector<Reference>({"kernel.out"}));

    auto pi3 = peer_info3();
    ASSERT_EQ(pi3.get_peer_ports("out"), std::vector<Reference>({"other.in"}));
    ASSERT_EQ(pi3.get_peer_ports("in"), std::vector<Reference>({"other.out"}));
}

TEST(libmuscle_peer_info, get_peer_dims) {
    auto pi = peer_info();
    ASSERT_EQ(pi.get_peer_dims("other"), std::vector<int>({}));

    auto pi2 = peer_info2();
    ASSERT_EQ(pi2.get_peer_dims("kernel"), std::vector<int>({20}));

    auto pi3 = peer_info3();
    ASSERT_EQ(pi3.get_peer_dims("other"), std::vector<int>({}));
}

TEST(libmuscle_peer_info, get_peer_locations) {
    auto pi = peer_info();
    ASSERT_EQ(pi.get_peer_locations("other"),
            std::vector<std::string>({"tcp:other"}));

    auto pi2 = peer_info2();
    ASSERT_EQ(pi2.get_peer_locations("kernel"),
            std::vector<std::string>({"tcp:kernel"}));

    auto pi3 = peer_info3();
    ASSERT_EQ(pi3.get_peer_locations("other"),
            std::vector<std::string>({"tcp:other"}));
}

TEST(libmuscle_peer_info, get_peer_endpoint) {
    auto pi = peer_info();
    ASSERT_EQ(std::string(pi.get_peer_endpoints("out", {})[0]), "other.in[13]");
    ASSERT_EQ(std::string(pi.get_peer_endpoints("in", {})[0]), "other.out[13]");

    auto pi2 = peer_info2();
    ASSERT_EQ(std::string(pi2.get_peer_endpoints("out", {11})[0]), "kernel[11].in");
    ASSERT_EQ(std::string(pi2.get_peer_endpoints("in", {11})[0]), "kernel[11].out");

    auto pi3 = peer_info3();
    ASSERT_EQ(std::string(pi3.get_peer_endpoints("out", {42})[0]), "other.in[42]");
    ASSERT_EQ(std::string(pi3.get_peer_endpoints("in", {42})[0]), "other.out[42]");
}

File: libmuscle/python/libmuscle/test/test_peer_info.py
import pytest

from ymmsl import Conduit, Identifier as Id, Reference as Ref

from libmuscle.peer_info import PeerInfo


@pytest.fixture
def peer_info() -> PeerInfo:
    kernel = Ref('kernel')
    index = [13]

    conduits = [
            Conduit('kernel.out', 'other.in'),
            Conduit('other.out', 'kernel.in')]

    peer_dims = {
            Ref('other'): []}

    peer_locations = {
            Ref('other'): ['tcp:other']}

    return PeerInfo(kernel, index, conduits, peer_dims, peer_locations)


@pytest.fixture
def peer_info2() -> PeerInfo:
    kernel = Ref('other')
    index = []

    conduits = [
            Conduit('kernel.out', 'other.in'),
            Conduit('other.out', 'kernel.in')]

    peer_dims = {
            Ref('kernel'): [20]}

    peer_locations = {
            Ref('kernel'): ['tcp:kernel']}

    return PeerInfo(kernel, index, conduits, peer_dims, peer_locations)


@pytest.fixture
def peer_info3() -> PeerInfo:
    kernel = Ref('kernel')
    index = []

    conduits = [
            Conduit('kernel.out', 'other.in'),
            Conduit('other.out', 'kernel.in')]

    peer_dims = {
            Ref('other'): []}

    peer_locations = {
            Ref('other'): ['tcp:other']}

    return PeerInfo(kernel, index, conduits, peer_dims, peer_locations)


def test_create_peer_info(peer_info) -> None:
    assert True


def test_is_connected(peer_info) -> None:
    assert peer_info.is_connected(Id('out'))
    assert peer_info.is_connected(Id('in'))
    assert not peer_info.is_connected(Id('not_connected'))


def test_get_peer_port(peer_info, peer_info2, peer_info3) -> None:
    assert peer_info.get_peer_ports(Id('out')) == [Ref('other.in')]
    assert peer_info.get_peer_ports(Id('in')) == [Ref('other.out')]

    assert peer_info2.get_peer_ports(Id('out')) == [Ref('kernel.in')]
    assert peer_info2.get_peer_ports(Id('in')) == [Ref('kernel.out')]

    assert peer_info3.get_peer_ports(Id('out')) == [Ref('other.in')]
    assert peer_info3.get_peer_ports(Id('in')) == [Ref('other.out')]


def test_get_peer_dims(peer_info, peer_info2, peer_info3) -> None:
    assert peer_info.get_peer_dims(Ref('other')) == []
    assert peer_info2.get_peer_dims(Ref('kernel')) == [20]
    assert peer_info3.get_peer_dims(Ref('other')) == []


def test_get_peer_locations(peer_info, peer_info2, peer_info3) -> None:
    assert peer_info.get_peer_locations(Ref('other')) == ['tcp:other']
    assert peer_info2.get_peer_locations(Ref('kernel')) == ['tcp:kernel']
    assert peer_info3.get_peer_locations(Ref('other')) == ['tcp:other']


def test_get_peer_endpoint(peer_info, peer_info2, peer_info3) -> None:
    assert str(peer_info.get_peer_endpoints(Id('out'), [])[0]) == 'other.in[13]'
    assert str(peer_info.get_peer_endpoints(Id('in'), [])[0]) == 'other.out[13]'

    assert str(peer_info2.get_peer_endpoints(Id('out'), [11])[0]) == 'kernel[11].in'
    assert str(peer_info2.get_peer_endpoints(Id('in'), [11])[0]) == 'kernel[11].out'

    assert str(peer_info3.get_peer_endpoints(Id('out'), [42])[0]) == 'other.in[42]'
    assert str(peer_info3.get_peer_endpoints(Id('in'), [42])[0]) == 'other.out[42]'
